File: 2905-find-indices-with-index-and-value-difference-ii/2905-find-indices-with-index-and-value-difference-ii.py
class Solution:
    def findIndices(self, nums: List[int], indexDifference: int, valueDifference: int) -> List[int]:
        min_idx = 0
        max_idx = 0
        
        for i in range(indexDifference, len(nums)):
            j = i - indexDifference

            if nums[j] < nums[min_idx]:
                min_idx = j
            if nums[j] > nums[max_idx]:
                max_idx = j
            
            if nums[max_idx] - nums[i] >= valueDifference:
                return [max_idx, i]
            if nums[i] - nums[min_idx] >= valueDifference:
                return [min_idx, i]
        
        return [-1, -1]

File: 2905-find-indices-with-index-and-value-difference-ii/2905-find-indices-with-index-and-value-difference-ii.java
class Solution {
    public int[] findIndices(int[] nums, int indexDifference, int valueDifference) {
        int mini = 0, maxi = 0;
        int n = nums.length;
        for (int i = 0; i < n; i++) {
            if (i - maxi >= indexDifference && Math.abs(nums[i] - nums[maxi]) >= valueDifference) {
                return new int[] {maxi, i};
            } else if (i - mini >= indexDifference && Math.abs(nums[i] - nums[mini]) >= valueDifference) {
                return new int[] {mini, i};
            }
            if (nums[i] > nums[maxi]) maxi = i;
            else if (nums[i] < nums[mini]) mini = i;
        }
        return new int[] {-1, -1};
    }
}
